storage: prevent deadlocking recursive heartbeats in mtc.expireLeases

andreimatei opened the pull request:
TestRefreshPendingCommands was flaky because, when done in the context
of a multiTestContext, a liveness heartbeat could be recursive (see bug
for details) and we don't allow more than one heartbeat at a time. This
stems from the multiTestContext transport hastily trying to increment
all the epochs on certain errors.
The solution was to get rid of these increments. This changes the
semantics of mtc.expireLeases(), which has been renamed to
mtc.advanceClock().

Fixes #13199

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13354)
<!-- Reviewable:end -->

andreimatei committed:
storage: prevent deadlocking recursive heartbeats in mtc.expireLeases

TestRefreshPendingCommands was flaky because, when done in the context
of a multiTestContext, a liveness heartbeat could be recursive (see bug
for details) and we don't allow more than one heartbeat at a time. This
stems from the multiTestContext transport hastily trying to increment
all the epochs on certain errors, and incrementing the epoch implies a
heartbeat) => deadlock.
The solution was to get rid of these increments. This changes the
semantics of mtc.expireLeases(), which has been renamed to
mtc.advanceClock().

A change was required to mtc.transferLease() so that we don't transfer
to a node whose liveness record is stale. This was a problem because, in
such cases, the transfer was inconsequential and we had tests that did a
transfer and then tried to unreplicate from the old leaseholder.
Unreplication was failing because, in the process of unreplicating, the
replica was acquiring a new lease, which then prohibited the
ChangeReplicas (one can't remove the leaseholder).

Fixes #13199

andreimatei committed:
util: refact WaitGroup.Wait() to return the error

andreimatei commented:
Sorry, I  messed up something here. Hold off on the review.

andreimatei commented:
OK now it's ready for review.

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-:-Kc01JI8AbBL1BAAygeM:b-l7h7xa)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r1, 5 of 7 files at r2.
Review status: 8 of 10 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_raft_test.go, line 2794 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc05OgU3dkeK3TZ0EaH:-Kc05OgU3dkeK3TZ0EaI:blpxf4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_raft_test.go#L2794)):*
> ```Go
> 	// (after the simulated passage of time).
> 	mtc.advanceClock(context.TODO())
> 	mtc.manualClock.Increment(int64(storage.ReplicaGCQueueInactivityThreshold + 1))
> ```

this pattern of mtc.advanceClock followed by mtc.manualClock.Increment seems weird; perhaps `advanceClock` is just poorly named?

---

*[pkg/storage/client_test.go, line 214 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc04spMYEAfkLhZEgZQ:-Kc04spMYEAfkLhZEgZR:bu4a695) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_test.go#L214)):*
> ```Go
> 	stoppers       []*stop.Stopper
> 	idents         []roachpb.StoreIdent
> 	nodeLivenesses []*storage.NodeLiveness
> ```

why this move?

---

*[pkg/util/waitgroup.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc046BIz5wiNL2428HS:-Kc046BIz5wiNL2428HT:bgyzlv8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddf10624979487ec9411bef2d67ef0535571f5b5/pkg/util/waitgroup.go#L15)):*
> ```Go
> // permissions and limitations under the License.
> //
> // Author: Dan Harrison (daniel.harrison@gmail.com)
> ```

Feel free to remove this

---

*[pkg/util/waitgroup_test.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc049OSWVSRnAB9o1fg:-Kc049OSWVSRnAB9o1fh:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddf10624979487ec9411bef2d67ef0535571f5b5/pkg/util/waitgroup_test.go#L15)):*
> ```Go
> // permissions and limitations under the License.
> //
> // Author: Dan Harrison (daniel.harrison@gmail.com)
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 8 of 10 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_raft_test.go, line 2794 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc05OgU3dkeK3TZ0EaH:-Kc0AJXiD0s5s0XBt6AD:b-52t6oq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_raft_test.go#L2794)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this pattern of mtc.advanceClock followed by mtc.manualClock.Increment seems weird; perhaps `advanceClock` is just poorly named?
</blockquote></details>

this pattern is not ideal :)
But I'm not sure what a better name for `advanceClock` would be... Cause it does a pretty weird thing (advances the clock by 2 * lease duration).

---

*[pkg/storage/client_test.go, line 214 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc04spMYEAfkLhZEgZQ:-Kc09h0YtFM3UnhmCVX9:b-ux44sw) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_test.go#L214)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why this move?
</blockquote></details>

cause `nodeLivenesses` is protected by `mu`

---

*[pkg/util/waitgroup.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc046BIz5wiNL2428HS:-Kc0A0Blq55zwaid69-7:bjdgj30) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddf10624979487ec9411bef2d67ef0535571f5b5/pkg/util/waitgroup.go#L15)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Feel free to remove this
</blockquote></details>

remove the author? Why? I thought that maybe we didn't use to do it in the ccl repo, but we do do it over here. @danhhz?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 7 files at r2.
Review status: 9 of 10 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_raft_test.go, line 2794 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc05OgU3dkeK3TZ0EaH:-Kc0rL98BEsFj_FQoNZu:bewlqv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_raft_test.go#L2794)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this pattern is not ideal :)
But I'm not sure what a better name for `advanceClock` would be... Cause it does a pretty weird thing (advances the clock by 2 * lease duration).
</blockquote></details>

But why do you need both? `mtc.advanceClock` just looks for the first store's lease expiration; all you want here is to advance the clock by the greater of the two durations.

This pattern is also all over the place, seems worth extracting a function.

---

*[pkg/storage/client_test.go, line 1121 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc0qlKgbD6dijESOOcU:-Kc0qlKgbD6dijESOOcV:bs1gb8b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_test.go#L1121)):*
> making this transfer irrelevant.

How does the transfer become irrelevant? Actually, how come it's OK to ping the liveness key _after_ the transfer?

---

*[pkg/storage/client_test.go, line 1138 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc0qzfrUT9cFNuDar6u:-Kc0qzfrUT9cFNuDar6v:b-5difgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_test.go#L1138)):*
> The liveness records all the nodes

accidentally a word

---

*[pkg/storage/client_test.go, line 1141 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc0r_G1AsBiAZg3erf6:-Kc0r_G1AsBiAZg3erf7:b8a2f1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_test.go#L1141)):*
> ```Go
> // on the new clock value (and this will cause all the epoch-based leases to be
> // considered expired until the liveness record is heartbeated).
> func (m *multiTestContext) advanceClock(ctx context.Context) {
> ```

this context is unused

---

*[pkg/storage/client_test.go, line 1150 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc0s48GpJxWPNRgzPtC:-Kc0s48GpJxWPNRgzPtD:brlw781) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_test.go#L1150)):*
> ```Go
> 	for _, store := range m.stores {
> 		if store != nil {
> 			m.manualClock.Increment(store.LeaseExpiration(m.clock))
> ```

not really your code, but why can't this be:
```
m.manualClock.Increment(2 * (m.storeConfig.RangeLeaseActiveDuration+m.clock.MaxOffset()).Nanoseconds())
```

---

*[pkg/util/waitgroup.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc046BIz5wiNL2428HS:-Kc0rG6-Ww_Uk5bthKSH:b-7dkdco) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddf10624979487ec9411bef2d67ef0535571f5b5/pkg/util/waitgroup.go#L15)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

remove the author? Why? I thought that maybe we didn't use to do it in the ccl repo, but we do do it over here. @danhhz?
</blockquote></details>

It appears to be done by personal preference.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Review status: 9 of 10 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/util/waitgroup.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc046BIz5wiNL2428HS:-Kc3WJxXQXRQe9ZEguh5:b5xj36f) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddf10624979487ec9411bef2d67ef0535571f5b5/pkg/util/waitgroup.go#L15)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

It appears to be done by personal preference.
</blockquote></details>

I don't care either way

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 3 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/storage/client_raft_test.go, line 2794 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc05OgU3dkeK3TZ0EaH:-Kc49P5BJFkvc7n4vOG4:bdy5dlb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_raft_test.go#L2794)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

But why do you need both? `mtc.advanceClock` just looks for the first store's lease expiration; all you want here is to advance the clock by the greater of the two durations.

This pattern is also all over the place, seems worth extracting a function.
</blockquote></details>

These two things do seem indeed to be done together a few times (but not always, there are tests that just advance by `ReplicaGCQInactivity`), but I'll let someone else consolidate it. The two things are conceptually different (one deals with leases, the other with GC-ing), I'd let tests deal with them explicitly (and also I don't really want to understand what all these tests are doing and whether they really need both).

---

*[pkg/storage/client_test.go, line 1121 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc0qlKgbD6dijESOOcU:-Kc4AQgc0Esh005k3YbA:b-14vj8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_test.go#L1121)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> making this transfer irrelevant.

How does the transfer become irrelevant? Actually, how come it's OK to ping the liveness key _after_ the transfer?

</blockquote></details>

the transfer would be irrelevant because an epoch-based lease that is transferred to a node with an expired liveness record doesn't matter. Any node can immediately take that lease.
It's OK to do the ping because, well, why wouldn't it be? The liveness table is in a range generally different than the range that we just transferred. And also we do the heartbeating through the destination node, so, if we just transferred the liveness records' range itself, the lease shouldn't move from `dest`.

---

*[pkg/storage/client_test.go, line 1138 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc0qzfrUT9cFNuDar6u:-Kc4BZz5aSmBYpSp5IGN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_test.go#L1138)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> The liveness records all the nodes

accidentally a word
</blockquote></details>

Done.

---

*[pkg/storage/client_test.go, line 1141 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc0r_G1AsBiAZg3erf6:-Kc4Br_ClzvpXU3XUw1F:b1gjqrl) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_test.go#L1141)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this context is unused
</blockquote></details>

hmmm why didn't the linter yell at me?
I've left it though, renamed to `_`.

---

*[pkg/storage/client_test.go, line 1150 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc0s48GpJxWPNRgzPtC:-Kc4CLrUMJfcRvx7owOK:b-ajv0o5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_test.go#L1150)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

not really your code, but why can't this be:
```
m.manualClock.Increment(2 * (m.storeConfig.RangeLeaseActiveDuration+m.clock.MaxOffset()).Nanoseconds())
```
</blockquote></details>

well the `store.LeaseExpiration()` function is used in a couple of places...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 7 files at r3, 6 of 7 files at r4.
Review status: 9 of 10 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_test.go, line 1121 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc0qlKgbD6dijESOOcU:-Kc4J7IURVbQI-RIbFTO:b-4hvzdj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_test.go#L1121)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

the transfer would be irrelevant because an epoch-based lease that is transferred to a node with an expired liveness record doesn't matter. Any node can immediately take that lease.
It's OK to do the ping because, well, why wouldn't it be? The liveness table is in a range generally different than the range that we just transferred. And also we do the heartbeating through the destination node, so, if we just transferred the liveness records' range itself, the lease shouldn't move from `dest`.
</blockquote></details>

You didn't answer my question, so I'll increase the emphasis:

How come it's OK to ping the liveness key **after** the transfer?

---

*[pkg/storage/client_test.go, line 1150 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc0s48GpJxWPNRgzPtC:-Kc4L2kMOe6O2B1PN0hl:bu4u5e5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_test.go#L1150)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

well the `store.LeaseExpiration()` function is used in a couple of places...
</blockquote></details>

That doesn't answer my question; the point is that you can skip the iteration here.

---

*[pkg/storage/client_test.go, line 1140 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc4JQqsGzkcTPdZOyeB:-Kc4JQqsGzkcTPdZOyeC:bc8jg7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbda936cb61af08b8e94c16c0d474bdd63c9f691/pkg/storage/client_test.go#L1140)):*
> ```Go
> // expiration-based leases become expired. The liveness records of all the nodes
> // will also become expired on the new clock value (and this will cause all the
> // epoch-based leases to be considered expired until the liveness record are
> ```

why did you change this `is` to `are`?

---

*[pkg/storage/client_test.go, line 1144 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc4Kh_dYXbvlpJAimTv:-Kc4Kh_e4MymGckpX1fS:b-mcs8ft) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbda936cb61af08b8e94c16c0d474bdd63c9f691/pkg/storage/client_test.go#L1144)):*
> ```Go
> //
> // This method should only be used when all the stores share the manual clock.
> // Otherwise, the desired effect would be ambiguous.
> ```

you have a panic in here now.

---

*[pkg/storage/client_test.go, line 1153 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc4KjkJ-1Ixe_36HLLl:-Kc4KjkJ-1Ixe_36HLLm:b-om593o) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbda936cb61af08b8e94c16c0d474bdd63c9f691/pkg/storage/client_test.go#L1153)):*
> ```Go
> 	for _, clock := range m.clocks {
> 		if clock != m.clock {
> 			panic("advanceClock() can't be used when some of the stores are sharing " +
> ```

this error message is not particularly useful. how about
`panic(fmt.Sprinft("clock at index %d is different from the shared clock", i))`
you'll have to add `i` above, of course.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r1, 7 of 7 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 989 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc4c6uWz08_UlQG-gT8:-Kc4c6uXXkpdMDceKYI3:b-aw37nt) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbda936cb61af08b8e94c16c0d474bdd63c9f691/pkg/storage/client_raft_test.go#L989)):*
> ```Go
> 			mtc.advanceClock(context.Background())
> 
> 			wg.Wait()
> ```

I didn't notice this when this class was introduced, but `WaitGroupWithError.Wait` should probably return an `error`. 

---

*[pkg/storage/client_test.go, line 1121 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc0qlKgbD6dijESOOcU:-Kc4_N2FEepzzadeFtsI:b3tsii) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_test.go#L1121)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

You didn't answer my question, so I'll increase the emphasis:

How come it's OK to ping the liveness key **after** the transfer?
</blockquote></details>

+1. It looks to me like it would be better to do it before; there might be some background process trying to steal the lease at the same time.

---

*[pkg/storage/client_test.go, line 1150 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc0s48GpJxWPNRgzPtC:-Kc4a49W3NjmXSu643He:b86mvf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_test.go#L1150)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

That doesn't answer my question; the point is that you can skip the iteration here.
</blockquote></details>

The idea is to avoid duplicating the LeaseExpiration logic here, but it would probably be better to make LeaseExpiration a method on the StoreConfig instead of the Store.

---

*[pkg/storage/client_test.go, line 1145 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc4_bUBm6hJlNxg0dgX:-Kc4_bUBm6hJlNxg0dgY:b18f46u) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbda936cb61af08b8e94c16c0d474bdd63c9f691/pkg/storage/client_test.go#L1145)):*
> ```Go
> // This method should only be used when all the stores share the manual clock.
> // Otherwise, the desired effect would be ambiguous.
> func (m *multiTestContext) advanceClock(_ context.Context) {
> ```

Why take a context if it's unused? (especially since you're already touching every call site for the rename, and every one of those call sites is passing `context.TODO()`)

---

*[pkg/storage/client_test.go, line 1147 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc4aEYTQjtibbzEMBkL:-Kc4aEYTQjtibbzEMBkM:bb5g4fz) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbda936cb61af08b8e94c16c0d474bdd63c9f691/pkg/storage/client_test.go#L1147)):*
> ```Go
> func (m *multiTestContext) advanceClock(_ context.Context) {
> 	// Allow only one expiration in progress at a time.
> 	if !atomic.CompareAndSwapInt32(&m.advanceClockActive, 0, 1) {
> ```

This was present on `expireLeases` because that method could call itself reentrantly when it incremented the heartbeat. That's no longer possible, so you can remove this. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-:-Kc4cOq1fAcMmVWfjKeM:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
storage: move LeaseExpiration() from a store to a StoreConfig

... so that it can be done on a multiTestContext without worrying about
which stores are up and which are down.
This required changing the way in which defaults are assigned to the
StoreConfig (now a StoreConfig is always initialized with the default by
the ctor; it's no longer the responsability of the mtc), and the mtc has
to save a shared StoreConfig when it creates it.

andreimatei commented:
Review status: 0 of 12 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_raft_test.go, line 989 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc4c6uWz08_UlQG-gT8:-KcK874bQp3HoR0EDY0f:b-wnk86f) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbda936cb61af08b8e94c16c0d474bdd63c9f691/pkg/storage/client_raft_test.go#L989)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I didn't notice this when this class was introduced, but `WaitGroupWithError.Wait` should probably return an `error`. 
</blockquote></details>

Yeah I also noticed this. Added a commit getting rid of `FirstError()`.

---

*[pkg/storage/client_test.go, line 1150 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc0s48GpJxWPNRgzPtC:-KcKGygTG94R8pH-fUt6:bo3envo) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_test.go#L1150)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The idea is to avoid duplicating the LeaseExpiration logic here, but it would probably be better to make LeaseExpiration a method on the StoreConfig instead of the Store.
</blockquote></details>

Good idea. Added a commit doing that. But, as with everything in `mtc`,  it's not very clean due to the many ways in which configs in the `mtc` are initialized. PTAL.

---

*[pkg/storage/client_test.go, line 1140 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc4JQqsGzkcTPdZOyeB:-KcKBtlDbNoFWLv9mjGR:b8d6qbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbda936cb61af08b8e94c16c0d474bdd63c9f691/pkg/storage/client_test.go#L1140)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why did you change this `is` to `are`?
</blockquote></details>

bad leftover from an ill-fate attempt to make this method do something else.

---

*[pkg/storage/client_test.go, line 1144 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc4Kh_dYXbvlpJAimTv:-KcKAuKDHfhRcYxZEZBm:b-f9zca4) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbda936cb61af08b8e94c16c0d474bdd63c9f691/pkg/storage/client_test.go#L1144)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

you have a panic in here now.
</blockquote></details>

I've reworded the comment.

---

*[pkg/storage/client_test.go, line 1145 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc4_bUBm6hJlNxg0dgX:-KcKAXaw2fMsKNESbAC-:bw3hplp) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbda936cb61af08b8e94c16c0d474bdd63c9f691/pkg/storage/client_test.go#L1145)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why take a context if it's unused? (especially since you're already touching every call site for the rename, and every one of those call sites is passing `context.TODO()`)
</blockquote></details>

well particularly because I'm touching all callsites (so I'm introducing a `blame`), it's a good time to introduce this context. I frequently have to pass pipe it through various test functions while debugging.
I've added a log message to remove the complaint that it's not used :)

---

*[pkg/storage/client_test.go, line 1147 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc4aEYTQjtibbzEMBkL:-KcKBU9svcyRr4cgCswC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbda936cb61af08b8e94c16c0d474bdd63c9f691/pkg/storage/client_test.go#L1147)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This was present on `expireLeases` because that method could call itself reentrantly when it incremented the heartbeat. That's no longer possible, so you can remove this. 
</blockquote></details>

Done.

---

*[pkg/storage/client_test.go, line 1153 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc4KjkJ-1Ixe_36HLLl:-KcKBf2PYQS1c7BDgi6F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbda936cb61af08b8e94c16c0d474bdd63c9f691/pkg/storage/client_test.go#L1153)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this error message is not particularly useful. how about
`panic(fmt.Sprinft("clock at index %d is different from the shared clock", i))`
you'll have to add `i` above, of course.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 5 files at r1, 3 of 3 files at r6, 4 of 7 files at r7, 4 of 5 files at r8.
Review status: 11 of 12 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_raft_test.go, line 968 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-KcKe3ljEgZZhfHZxkYt:-KcKe3ljEgZZhfHZxkYu:behyi5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/c06397c005353ab27ec3900c48abb19c897852cc/pkg/storage/client_raft_test.go#L968)):*
> ```Go
> 				wg.Add(1)
> 				go func(i int) {
> 					if err := mtc.stores[i].DrainLeases(true); err != nil {
> ```

fun fact:
```
err := mtc.stores[i].DrainLeases(true)
wg.Done(errors.Wrapf(err, "store %d", i))
```

is equivalent

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 2 of 5 files at r1, 3 of 3 files at r6, 4 of 7 files at r7, 5 of 5 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/storage/client_test.go, line 1121 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354#-Kc0qlKgbD6dijESOOcU:-KcOipOXAMqfDTLM3xJh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ce7b7022b074195f4d1fb8e4183112aeed070f6/pkg/storage/client_test.go#L1121)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

+1. It looks to me like it would be better to do it before; there might be some background process trying to steal the lease at the same time.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13354)*
<!-- Sent from Reviewable.io -->
